File: mobile-engage/src/androidTest/java/com/emarsys/mobileengage/iam/InAppStartActionTest.java
package com.emarsys.mobileengage.iam;

import android.os.Handler;

import com.emarsys.mobileengage.MobileEngageInternal;
import com.emarsys.mobileengage.MobileEngageInternal_V3_Old;
import com.emarsys.testUtil.DatabaseTestUtils;
import com.emarsys.testUtil.SharedPrefsUtils;
import com.emarsys.testUtil.TimeoutUtils;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

public class InAppStartActionTest {

    private static final String EMARSYS_SHARED_PREFERENCES = "emarsys_shared_preferences";

    static {
        mock(Handler.class);
    }

    private MobileEngageInternal mobileEngageInternal;
    private InAppStartAction startAction;

    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();

    @Before
    @SuppressWarnings("unchecked")
    public void init() {
        DatabaseTestUtils.deleteCoreDatabase();
        SharedPrefsUtils.clearSharedPrefs(EMARSYS_SHARED_PREFERENCES);

        mobileEngageInternal = mock(MobileEngageInternal_V3_Old.class);

        startAction = new InAppStartAction(mobileEngageInternal);
    }

    @After
    public void tearDown() {
        DatabaseTestUtils.deleteCoreDatabase();
        DatabaseTestUtils.deleteCoreDatabase();
        SharedPrefsUtils.clearSharedPrefs(EMARSYS_SHARED_PREFERENCES);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_mobileEngageInternalMustNotBeNull() {
        new InAppStartAction(null);
    }

    @Test
    public void testExecute_callsMobileEngageInternal() {
        startAction.execute(null);

        verify(mobileEngageInternal).trackInternalCustomEvent("app:start", null, null);
    }

}

File: mobile-engage/src/androidTest/java/com/emarsys/mobileengage/iam/InAppStartActionTest.kt
package com.emarsys.mobileengage.iam

import com.emarsys.core.concurrency.CoreSdkHandlerProvider
import com.emarsys.core.di.DependencyInjection
import com.emarsys.core.storage.Storage
import com.emarsys.mobileengage.event.EventServiceInternal
import com.emarsys.mobileengage.fake.FakeMobileEngageDependencyContainer
import com.emarsys.mobileengage.util.waitForTask
import com.emarsys.testUtil.TimeoutUtils
import com.emarsys.testUtil.mockito.whenever
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.verifyZeroInteractions
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TestRule

class InAppStartActionTest {

    private lateinit var mockEventServiceInternal: EventServiceInternal
    private lateinit var mockContactTokenStorage: Storage<String?>
    private lateinit var startAction: InAppStartAction

    @Rule
    @JvmField
    val timeout: TestRule = TimeoutUtils.timeoutRule

    @Before
    fun setUp() {
        mockEventServiceInternal = mock()
        mockContactTokenStorage = mock()

        DependencyInjection.setup(FakeMobileEngageDependencyContainer(coreSdkHandler = CoreSdkHandlerProvider().provideHandler()))

        startAction = InAppStartAction(mockEventServiceInternal, mockContactTokenStorage)
    }

    @After
    fun tearDown() {
        DependencyInjection.tearDown()
    }

    @Test
    fun testExecute_callsEventServiceInternal_whenContactTokenIsPresent() {
        whenever(mockContactTokenStorage.get()).thenReturn("contactToken")

        startAction.execute(null)
        waitForTask()

        verify(mockEventServiceInternal).trackInternalCustomEvent("app:start", null, null)
    }

    @Test
    fun testExecute_EventServiceInternal_shouldNotBeCalled_whenContactTokenIsNotPresent() {
        whenever(mockContactTokenStorage.get()).thenReturn(null)

        startAction.execute(null)
        waitForTask()

        verifyZeroInteractions(mockEventServiceInternal)
    }
}